Add General Permissions graph APIs

qnkhuat opened the pull request:
Epic: https://github.com/metabase/metabase/issues/20397
Product's doc: https://www.notion.so/metabase/Add-feature-level-permissions-to-groups-8027685681774932ac7221131f061295

This PR is part of a series of PRs to add support for general permissions. Specifically this PR adds:
- `GET /api/ee/advanced-permissions/general/graph` - used to get General Permissions graph
- `PUT /api/ee/advanced-permissions/general/graph` - used to update General Permissions graphs

Example graph:
```json
{
    "revision": 1,
    "groups": {
        "8": {
            "setting": "yes",
            "monitoring": "yes",
            "subscription": "no"
        },
        "2": {
            "setting": "yes",
            "monitoring": "yes",
            "subscription": "yes"
        }
    }
}
 ```

Things that will come later so General permissions can be fully functional
- Migration to enable Subscription permission by default
- Modify `/api/user/current` to return general permissions info so FE could decide when to display a certain features
- Enable users with these permissions could access related APIs

qnkhuat committed:
add API to fetch general permisisons graph

qnkhuat committed:
add API to update general permissionns

qnkhuat committed:
change author of migration

qnkhuat committed:
merge with master to resolve conflict in migration

qnkhuat committed:
update documents

qnkhuat committed:
misc fixes to applease the CIs

qnkhuat committed:
Add tests for general permission APIs and models

qnkhuat committed:
linting and fix a failed test case

qnkhuat committed:
fix some failed tests

qnkhuat committed:
update docs and change /subscription/ to /general/subscription/ for consistency

qnkhuat commented on enterprise/backend/src/metabase_enterprise/advanced_permissions/models/permissions/general_permissions.clj:
```diff
@@ -0,0 +1,126 @@
+(ns metabase-enterprise.advanced-permissions.models.permissions.general-permissions
```
This works pretty much like `metabase.models.collection.graph`

qnkhuat commented on resources/migrations/000_migrations.yaml:
```diff
@@ -11246,6 +11246,54 @@ databaseChangeLog:
         - sql:
             sql: DELETE FROM permissions_group WHERE name = 'MetaBot'
 
+  - changeSet:
```
this new table is very much like `CollectionPermissionRevision` or `PermissionRevision` with the only difference that we don't have the `after` column in this table, instead, we have `changes`.

Note: both `CollectionPermissionRevision` and `PermissionRevision` have `after` column but it actually store the `changes` not the `after` graph.

qnkhuat commented on src/metabase/models/permissions.clj:
```diff
@@ -255,6 +255,15 @@
+               ;; /general/setting/    -> permissions to access /admin/settings page
+               ;; /general/monitoring/ -> permissions to access tools, audit and troubleshooting
+               ;; /general/subscription/ -> permisisons to create/edit subscriptions and alerts
+               (and "general/"
```
Even though we could ignore the `/general/` prefix and it'll work just fine, but I'm still adding it to make the path clearer bc we might have more `general` permissions later on.

noahmoss commented on enterprise/backend/src/metabase_enterprise/advanced_permissions/api/general.clj:
```diff
@@ -0,0 +1,41 @@
+
+(api/defendpoint GET "/graph"
+  "Fetch a graph of General Permissions"
+  [:as _req]
```
nit: can just leave the arg list empty

noahmoss commented on enterprise/backend/src/metabase_enterprise/advanced_permissions/models/permissions/general_permissions.clj:
```diff
@@ -0,0 +1,126 @@
+                                       {:where [:or
+                                                [:= :object "/"]
+                                                [:like :object (hx/literal "/general/%")]]})
+        permissions-set-from-group-id (fn [group-id]
+                                        (->> general-permissions
+                                             (filter #(= (:group_id %) group-id))
+                                             (map :object)
+                                             set))]
+    (into {} (for [group-id group-ids]
+               {group-id (permissions-set-from-group-id group-id)}))))
```
Can't you do this with `group-by`, like the implementation of the function of the same name in `metabase.models.collection.graph`?

qnkhuat replied:
No, the reason `metabase.models.collection.graph/group-id->permissions-set` will return a map of all group id that has `permissions`, That map might not contain a `group` if that group has no permissions. 

With my implentation, it's guaranteed that the function will return a map for all Permission groups, whether a Permission Group has permission or not

noahmoss replied:
that's fine, though, right? if the group doesn't show up in the graph, that just means it has no permissions. 

we don't enforce that the data perms graph has every group in it, for example

qnkhuat replied:
then I think it's the question where do get the source of all `groups`, is it the graph itself, or from somewhere else?

maybe @alxnddr could help me answer this?
The question is: should a group with no general permissions show up in the graph?

My answer is it should.

alxnddr replied:
@qnkhuat both options work but the approach Noah described is already implemented for data and collection permission graphs. It also helps to transfer fewer data between BE and FE. What do you think about sticking to this approach?

qnkhuat replied:
if you're okay with it then I'll go with what Noah described. 
I was worried that if the graph doesn't show a group, then FE will not aware of it, but seems like it's not the case

noahmoss commented on test/metabase/test/util.clj:
```diff
@@ -264,6 +264,11 @@
   [obj]
   (json/parse-string (json/generate-string obj) keyword))
 
+(defn int->kw
+  "Convert an integer to a keyword. I.e : 3 -> :3"
+  [x]
+  (keyword (str x)))
```
Hmm...having to create keywords out of ints is a code smell, but it seems like this is a side effect of using `keyword` in the test HTTP client to [parse all JSON keys](https://github.com/metabase/metabase/blob/60bfbc61b94df6915ea6fd73f83046d08cf5fa7c/test/metabase/http_client.clj#L85). It hasn't been an issue before because we just don't have API-level tests for the other graph fetching endpoints.

Maybe you could try changing that HTTP client parsing code to parse ints properly in the response, rather than indiscriminately making them keywords? If that works and doesn't break a ton of other tests, it's a more future-proof solution.

qnkhuat replied:
Nice catch, but I lets do it in a another PR, I tried to parse it but we have 13 failing tests.
[Change](https://github.com/metabase/metabase/pull/21124/commits/da4105947fad41ae43293b9a33ed94dcb3df76b2)
[CI](https://app.circleci.com/pipelines/github/metabase/metabase/31543/workflows/2ec0220f-9552-437e-90bf-d35577edd4c9/jobs/1599343)

noahmoss reviewed (changes requested):
Looks mostly good, but have some comments & questions. Will do another pass once these are addressed.

qnkhuat commented on enterprise/backend/src/metabase_enterprise/advanced_permissions/models/permissions/general_permissions.clj:
```diff
@@ -0,0 +1,126 @@
+(def ^:private GeneralPermissions (s/enum :yes :no))
+
+(def ^:private GroupPermissionsGraph
+  {:setting      GeneralPermissions
+   :monitoring   GeneralPermissions
+   :subscription GeneralPermissions
+   s/Keyword     GeneralPermissions})
```
nice 👍

qnkhuat commented on enterprise/backend/src/metabase_enterprise/advanced_permissions/models/permissions/general_permissions.clj:
```diff
@@ -0,0 +1,126 @@
+
+;;; -------------------------------------------------- Update Graph --------------------------------------------------
+
+(defn- save-perms-revision!
```
for the 2 others revision table, I think yes, but for my `GeneralPermissionsRevision`, I don't use the `after` column, instead I use `changes` column as noted here: https://github.com/metabase/metabase/pull/21124#discussion_r832052628.

If we think going with `after` for consistency is better, then we can generalize all of this. Wdyt?

noahmoss replied:
oh, I see. I agree that the name `changes` makes more sense given what we're storing in that row. but in this case, my feeling is consistently might be best.

in theory you could write a migration to copy the value in `after` to `changes` for both the existing tables, and then in a future release, delete the `after` column. up to you if you want to tackle that now

qnkhuat replied:
make sense, but let's me tackle this in a seperate PR : https://github.com/metabase/metabase/pull/21270

qnkhuat replied:
are there any reason we need to copy value and delete later? I think we could simply rename the column right?

qnkhuat commented on enterprise/backend/src/metabase_enterprise/advanced_permissions/models/permissions/general_permissions.clj:
```diff
@@ -0,0 +1,126 @@
+  (let [old-graph          (graph)
+        old-perms          (:groups old-graph)
+        new-perms          (:groups new-graph)
+        ;; filter out any groups not in the old graph
+        new-perms          (select-keys new-perms (keys old-perms))
+        ;; filter out any collections not in the old graph
+        new-perms          (into {} (for [[group-id collection-id->perms] new-perms]
+                                      [group-id (select-keys collection-id->perms (keys (get old-perms group-id)))]))
```
ah, I didn't change the name. but I think the gist is still valid for general permisison, because it's to prevent:
- accidiently grant a new permission that we don't know about
- grant for permisisons a new group

noahmoss replied:
wouldn't a group with no general permissions not show up in the old graph? but we still want to be able to set permissions for it

and collections shouldn't be relevant here at all, right? so I don't get what the last line is doing

qnkhuat replied:
ah, it actually relates to our other comment about how to get group-ids->permisison-set.
Currently if a group has no general permissions at all, we still have in the graph

qnkhuat committed:
address noah's comments

qnkhuat committed:
Parse number as is in http-client for tests

qnkhuat committed:
revert the last commit about parsing API response in tests

qnkhuat committed:
Merge branch 'master' into ngoc-general-permission

qnkhuat committed:
change fk name

qnkhuat committed:
Changes:
- Rename `changes` column to `after` to keep things consistent
- If a group  has no General Permisions, it'll not be included in the graph
- Update tests and some docs

qnkhuat committed:
fix failing tests in ee

qnkhuat committed:
add some tests and make docstring completes

noahmoss commented on src/metabase/models/permissions.clj:
```diff
@@ -255,6 +258,15 @@
                                (and "schema/"
                                     (opt (and path-char "*/"
                                               (opt #"table/\d+/"))))))))
+               ;; any path starting with /general is a permission to access certain features of admin settings page
```
nit: subscriptions are not a feature in the admin panel. maybe say something like `any path starting with /general is a permission that is not scoped by database or collection`? or something to that effect

noahmoss reviewed (approved):
lgtm! a few small comments but nothing blocking. I've also merged https://github.com/metabase/metabase/pull/21216 so you'll have to modify your tests accordingly

qnkhuat commented on enterprise/backend/test/metabase_enterprise/advanced_permissions/api/general_test.clj:
```diff
@@ -0,0 +1,68 @@
+            new-graph     (assoc-in current-graph [:groups (mt/int->kw group-id) :setting] "yes")]
+
+        (premium-features-test/with-premium-features #{}
+          (testing "Should require a token with `:advanced-permissions`"
```
yea :) it's a bit misleading when reading from a test title POV. I'll change it

qnkhuat replied:
I think if you read the full test title it will make more sense,  it'll be like this `PUT /api/ee/advanced-permissions/general/graph Should require a token with :advanced-permissions`

Basically saying this API will require a special feature to work.




FYI we use this title title in various place to check premium features:
https://github.com/metabase/metabase/blob/f6844db1a1779910ca5b89b9d22f2f215de2bc33/enterprise/backend/test/metabase_enterprise/audit_app/api/user_test.clj#L10

qnkhuat commented on enterprise/backend/test/metabase_enterprise/advanced_permissions/api/general_test.clj:
```diff
@@ -0,0 +1,68 @@
+                 (mt/user-http-request :rasta :get 403 "ee/advanced-permissions/general/graph"))))
+
+        (testing "return general permissions for groups that has general permisions"
+          ;; TODO: remove this when subscription is granted by default for new group
+          (perms/grant-general-permissions! group-id :subscription)
```
agree with that comment, I'll change this comment to make it more accurate.

qnkhuat committed:
address Noah's comments and merge with master

qnkhuat committed:
fix namespaces
